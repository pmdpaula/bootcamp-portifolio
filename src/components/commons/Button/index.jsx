import get from 'lodash/get';
import { PropTypes } from 'prop-types';
import styled, { css } from 'styled-components';

import { TextStyleVariants } from '../../../foundation/Text';
import breakpointsMedia from '../../../theme/utils/breakpointsMedia';
import propToStyle from '../../../theme/utils/propToStyle';
import Link from '../Link/index';

const ButtonGhost = css`
  color: ${({ theme, variant }) => get(theme, `colors.${variant}.color`)};
  background-color: transparent;
`;

const ButtonDefault = css`
  color: ${({ theme, variant }) =>
    // eslint-disable-next-line implicit-arrow-linebreak
    get(theme, `colors.${variant}.contrastText`)};
  background-color: ${({ theme, variant }) =>
    // eslint-disable-next-line implicit-arrow-linebreak
    get(theme, `colors.${variant}.color`)};
`;

const ButtonWrapper = styled.button`
  border: 0;
  cursor: pointer;
  padding: 12px 26px;
  font-weight: bold;
  opacity: 1;
  transition: opacity ${({ theme }) => theme.transition};
  border-radius: ${({ theme }) => theme.defaultStyles.borderRadius};

  ${breakpointsMedia({
    xs: css`
      ${TextStyleVariants.smallestException}
    `,
    md: css`
      padding: 12px 43px;
      ${TextStyleVariants.paragraph1}
    `,
  })}

  ${({ ghost }) => (ghost ? ButtonGhost : ButtonDefault)}

  &:hover,
  &:focus {
    opacity: 0.5;
  }

  &:disabled {
    cursor: not-allowed;
    opacity: 0.2;
  }
  ${({ fullWidth }) =>
    fullWidth &&
    css`
      width: 100%;
    `};

  ${propToStyle('margin')}
  ${propToStyle('display')}
`;

const Button = ({ href, children, ...props }) => {
  const hasHref = Boolean(href);
  const tag = hasHref ? Link : 'button';

  return (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <ButtonWrapper as={tag} href={href} {...props}>
      {children}
    </ButtonWrapper>
  );
};

Button.defaultProps = {
  href: undefined,
};

Button.propTypes = {
  href: PropTypes.string,
  children: PropTypes.node.isRequired,
};

export default Button;
