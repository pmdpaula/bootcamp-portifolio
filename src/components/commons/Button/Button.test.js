import { render, screen } from '../../../infra/test/testUtils';
import Button from './index';

describe('<Button />', () => {
  // const { colors } = useContext(ThemeContext);
  test('renders component', () => {
    render(<Button href="#">Botão</Button>);

    // screen.debug();
    // expect(true).toBe(true);

    const button = screen.getByRole('link', { name: /botão/i });
    expect(button).toMatchSnapshot();
  });
});
