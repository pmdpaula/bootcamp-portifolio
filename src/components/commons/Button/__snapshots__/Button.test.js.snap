// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders component 1`] = `
.c0 {
  color: inherit;
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  opacity: 1;
  -webkit-transition: opacity;
  transition: opacity;
}

.c0:hover,
.c0:focus {
  opacity: 0.5;
}

.c1 {
  border: 0;
  cursor: pointer;
  padding: 12px 26px;
  font-weight: bold;
  opacity: 1;
  -webkit-transition: opacity;
  transition: opacity;
  border-radius: 6px;
}

.c1:hover,
.c1:focus {
  opacity: 0.5;
}

.c1:disabled {
  cursor: not-allowed;
  opacity: 0.2;
}

@media only screen and (min-width:0px) {
  .c1 {
    font-size: 12px;
    font-weight: 400;
    line-height: 1;
  }
}

@media only screen and (min-width:768px) {
  .c1 {
    padding: 12px 43px;
    font-size: 16px;
    font-weight: 400;
    line-height: 1.25;
  }
}

<a
  class="c0 c1"
  href="/#"
>
  Bot√£o
</a>
`;
